### ISSUE TITLE
Unpacking unbounded tuple types do not recognise its sub type

### ISSUE THREAD

**Issue by airallergy** [python/mypy#17111]
**Bug Report**

If a function argument is type hinted using an unpacked unbounded tuple type, mypy complains when passing a sub type into that function.

**To Reproduce**

https://mypy-play.net/?mypy=1.9.0&python=3.11&gist=80a693ef86b0790477d714137d48ef0c

```python
def foo(x:tuple[int, *tuple[float|int,...]]):
    return x
    
y = (1, 2, 3)
foo(y)

z: tuple[int, ...]
foo(z)
```

**Expected Behavior**

mypy should be happy about both `foo(y)` and `foo(z)`.

**Actual Behavior**

mypy passes `foo(y)`, but raises `error: Argument 1 to "foo" has incompatible type "tuple[int, ...]"; expected "tuple[int, *tuple[float | int, ...]]"  [arg-type]` for  `foo(z)`.

**Your Environment**

- Mypy version used: 1.9.0
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.11

**Comment by erictraut** [comment]
The problem in this code is that the type `tuple[int, ...]` represents a tuple with zero or more elements, all of which are `int`, but the type `tuple[int, *tuple[float|int, ...]]` requires at least one element. A zero-length tuple is compatible with the first type but not the second. Mypy is correct to flag this as an error.

**Comment by airallergy** [comment]
Right, that makes sense, sorry about the false bug report. A follow-up question is, how could `z` be narrowed to have at least one element?

Intuitively, we could do `assert len(z) >= 1`, but that does not seem to work. Though `cast` works, would there be any alternatives to `cast`?

https://mypy-play.net/?mypy=1.9.0&python=3.11&gist=a5499828510394d72124fdd1cf12d0c7

```python
from typing import cast

def foo(x:tuple[int, *tuple[float|int,...]]):
    return x
    
def bar1(z: tuple[int, ...]):
    assert len(z) >= 1
    reveal_type(z)
    foo(z)
    
def bar2(z: tuple[int, ...]):
    z = cast(tuple[int, *tuple[int, ...]], z)
    reveal_type(z)
    foo(z)
```

**Comment by TeamSpen210** [comment]
Actually, mypy does know how to do this. But to preserve backwards compatibility, currently it only infers variadic tuples when code is already using them. If you turn on [`--enable-incomplete-feature=PreciseTupleTypes`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-enable-incomplete-feature), you can opt into inferring more precise types, which will make the assertion work correctly.

**Comment by airallergy** [comment]
Thanks, `--enable-incomplete-feature=PreciseTupleTypes` works like a charm, though it seems not available on the mypy playground.
